### PR TITLE
Add command-line "Go-To-Line" option

### LINKED ISSUE #415: Expose "Go To Line" feature externally
Expose the "go to line" feature so that it is accessible from another app or command line,

 ###  CommandLine Example 
 <pre><code> $ io.elementary.code [file_name] -[option] 15.13-15.31 </code> </pre>

Open [file_name] and highlight [from: line 15 character 13] [to: line 15 character 31].

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/57918621-expose-go-to-line-feature-externally?utm_campaign=plugin&utm_content=tracker%2F61917289&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F61917289&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by colinkiama**
Fixes #415

### CODE REVIEW

**Comment by jeremypw** [comment]
@colinkiama Just a hint - for a substantial PR, if it is ready for review then its best to request a review either from a specific person or from the elementary/desktop-developers team (if you are able to do that).

**Comment by jeremypw** [comment]
Just had a quick try of this - the commandline handling seems to work but the actual selecting and displaying the specified range does not seem to work yet? Maybe convert to draft if you are still working on this.

**Comment by colinkiama** [comment]
> @colinkiama Just a hint - for a substantial PR, if it is ready for review then its best to request a review either from a specific person or from the elementary/desktop-developers team (if you are able to do that).

I don't have the option to request a reviewer on this PR sadly:
Screenshot from 2023-11-26 15 12 19

**Comment by colinkiama** [comment]
> Just had a quick try of this - the command line handling seems to work but the actual selecting and displaying the specified range does not seem to work yet? Maybe convert to draft if you are still working on this.

I'm sure that I completed this. Do you mind sharing what you wrote in the terminal to try out this feature?

**Comment by jeremypw** [comment]
> > @colinkiama Just a hint - for a substantial PR, if it is ready for review then its best to request a review either from a specific person or from the elementary/desktop-developers team (if you are able to do that).
> 
> I don't have the option to request a reviewer on this PR sadly

Ah, I guess you have to be given that permission by @danirabbit . If you contribute regularly (which you seem to be doing) then that should not be a problem :smile:

**Comment by jeremypw** [comment]
> > Just had a quick try of this - the command line handling seems to work but the actual selecting and displaying the specified range does not seem to work yet? Maybe convert to draft if you are still working on this.
> 
> I'm sure that I completed this. Do you mind sharing what you wrote in the terminal to try out this feature?

Strange - just tried it again and it worked this time. Not sure what happened before - I'll do some more testing and review the coding.

File: src/Services/LocationJumpManager.vala [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+            return SelectionRange () {
+                start_line = parse_num_from_match_info (match_info, "start_line"),
+                end_line = parse_num_from_match_info (match_info, "end_line"),
+                start_column = parse_num_from_match_info (match_info, "start_column"),
+                end_column = parse_num_from_match_info (match_info, "end_column"),
+            };
+        }
+
+        private static int parse_num_from_match_info (MatchInfo match_info, string match_name) {
+            string str = match_info.fetch_named (match_name);
```
Comment (jeremypw) [inline src/Services/LocationJumpManager.vala:76]:
```suggestion
            var str = match_info.fetch_named (match_name);
```

File: src/Services/LocationJumpManager.vala [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            int num;
+
+            if (int.try_parse (str, out num)) {
```
Comment (jeremypw) [inline src/Services/LocationJumpManager.vala:79]: Theoretically `str` could be null as `fetch_named` returns `string?` according to https://valadoc.org/glib-2.0/GLib.MatchInfo.fetch_named.html.  But `try_parse` expects `string` according to https://valadoc.org/glib-2.0/int.try_parse.html so do we need to null check here?

File: src/Services/LocationJumpManager.vala [hunk 3]
```diff
@@ -0,0 +1,86 @@
+namespace Scratch {
```
Comment (jeremypw) [inline src/Services/LocationJumpManager.vala:1]: Header needed for this file.  The latest style of header we use is something like:
```
/*
 * SPDX-License-Identifier: GPL-3.0-or-later
 * SPDX-FileCopyrightText: 2023 elementary, Inc. <https://elementary.io>
 */
```
You may want to add an "Author" detail line as well.

File: src/Widgets/DocumentView.vala [hunk 4]
```diff
@@ -252,6 +252,54 @@ public class Scratch.Widgets.DocumentView : Granite.Widgets.DynamicNotebook {
...
+                if (range != SelectionRange.EMPTY) {
+                    doc.source_view.select_range (range);
+                }
+                save_opened_files ();
+            });
+
+            return false;
+        });
+    }
+
```
Comment (jeremypw) [inline src/Widgets/DocumentView.vala:302]: Can this function be combined with the original `open_document` function? There is a lot of repeated code here.

File: src/Services/LocationJumpManager.vala (hunk 2 above)
Comment (colinkiama) [reply src/Services/LocationJumpManager.vala:79]: While it seems to behave fine the way it is, I'll add a null check anyway just to be safe.

File: src/Widgets/DocumentView.vala (hunk 4 above)
Comment (colinkiama) [reply src/Widgets/DocumentView.vala]: Sure!

File: src/Services/LocationJumpManager.vala [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+        private static int parse_num_from_match_info (MatchInfo match_info, string match_name) {
+            var str = match_info.fetch_named (match_name);
+            int num = 0;
+
+            if (str != null && int.try_parse (str, out num)) {
+                return num;
+            }
+
+            return num;
```
Comment (jeremypw) [inline src/Services/LocationJumpManager.vala:90]:
```suggestion
            if (str != null) {
                int.try_parse (str, out num);
            }

            return num;
```

File: src/Application.vala [hunk 6]
```diff
@@ -143,15 +172,20 @@ namespace Scratch {
...
                     } else {
+                        debug ("Files length: %d\n", files.length);
                         var doc = new Scratch.Services.Document (window.actions, file);
-                        window.open_document (doc);
+                        if (location_jump_manager.has_selection_range != null && files.length == 1) {
+                            window.open_document_at_selected_range (doc, true, location_jump_manager.range);
+                        } else {
+                            window.open_document (doc);
+                        }
+
```
Comment (jeremypw) [inline src/Application.vala:188]:
```suggestion
```

File: src/Widgets/SourceView.vala [hunk 7]
```diff
@@ -476,6 +476,34 @@ namespace Scratch.Widgets {
             buffer.end_user_action ();
         }
 
+        public void select_range (SelectionRange range) {
+            if (range.start_line < 0) {
+                return;
+            }
+            Gtk.TextIter start_iter;
```
Comment (jeremypw) [inline src/Widgets/SourceView.vala:483]:
```suggestion
            }
 
            Gtk.TextIter start_iter;
```

**Review by jeremypw — CHANGES_REQUESTED** [review]
Just a couple of elementaryos code-style suggestions and one suggestion to avoid unnecessary multiple function returns.  Should be ready then.

File: src/Application.vala (hunk 6 above)
Comment (colinkiama) [reply src/Application.vala]: Done

File: src/Widgets/SourceView.vala (hunk 7 above)
Comment (colinkiama) [reply src/Widgets/SourceView.vala]: Done

**Review by jeremypw — APPROVED** [review]
Good to go now,  thanks for your work!

**Pull request merged**